app.js

```diff
@@ -108,43 +120,67 @@ router.post('/account/:phoneNumber/:accountId/requestCode', async ctx => {
+        const accessKeys = await nearAccount.getAccessKeys();
+        return accessKeys.some(it => {
+            const publicKey = it.public_key.replace('ed25519:', '');
+            return nacl.sign.detached.verify(hash, Buffer.from(signedData, 'base64'), bs58.decode(publicKey));
+        });
+    } catch (e) {
+        console.error(e);
+        return false;
     }
-    return accessKeys.some(it => {
-        const publicKey = it.public_key.replace('ed25519:', '');
-        return nacl.sign.detached.verify(hash, Buffer.from(signature, 'base64'), bs58.decode(publicKey));
-    });
 };
 
-// TODO: Different endpoints for setup and recovery
-router.post('/account/:phoneNumber/:accountId/validateCode', async ctx => {
-    const { phoneNumber, accountId } = ctx.params;
-    const { securityCode, signature, publicKey } = ctx.request.body;
+function recoveryMethodsFor(account) {
+    return {
+        email: account.email || null,
+        emailAddedAt: account.emailAddedAt || null,
+        phoneAddedAt: account.phoneAddedAt || null,
+        phoneNumber: account.phoneNumber || null,
+        phraseAddedAt: account.phraseAddedAt || null,
+    };
+}
+
+router.post('/account/recoveryMethods', checkAccountOwnership, async ctx => {
+    const { accountId } = ctx.request.body;
+    const account = await models.Account.findOne({ where: { accountId } });
+    ctx.body = recoveryMethodsFor(account || {});
+});
 
-    const account = await models.Account.findOne({ where: { accountId, phoneNumber } });
-    if (!account || !account.securityCode || account.securityCode != securityCode) {
-        ctx.throw(401);
+const recoveryMethodUpdaters = {
+    phone: account => account.update({ phoneNumber: null, phoneAddedAt: null }),
+    email: account => account.update({ email: null, emailAddedAt: null }),
+    phrase: account => account.update({ phraseAddedAt: null })
+};
+
+async function checkRecoveryMethod(ctx, next) {
+    const { recoveryMethod } = ctx.request.body;
+    const allRecoveryMethods = Object.keys(recoveryMethodUpdaters);
+    if (allRecoveryMethods.includes(recoveryMethod)) {
+        await next();
+        return;
     }
-    if (!account.confirmed) {
-        const nearAccount = await ctx.near.account(accountId);
-        const isSignatureValid = await verifySignature(nearAccount, securityCode, signature);
-        if (!isSignatureValid) {
-            ctx.throw(401);
-        }
-        await account.update({ securityCode: null, confirmed: true });
-    } else {
-        await (await ctx.near.account(accountId)).addKey(publicKey);
-        await account.update({ securityCode: null });
+    ctx.throw(400, `Given recoveryMethod '${recoveryMethod}' invalid; must be one of: ${allRecoveryMethods.join(', ')}`);
+}
+
+async function checkAccountInDB(ctx, next) {
+    const { accountId } = ctx.request.body;
+    const account = await models.Account.findOne({ where: { accountId } });
```

Comment:
any reason not to do it similarly to https://github.com/nearprotocol/near-fiddle-api/blob/c0ecec0b27fba1a0c961e7e0223c961ac113d50d/app.js#L60 so that `account` can be fetched from `ctx` instead of being queried from DB again?